go-controller/pkg/ovn/service.go

```diff
@@ -1,10 +1,140 @@
+	clusterServices := make(map[string][]string)
+
+	// For all nodePorts in k8s, we will populate the below slice with
+	// nodePort. In OVN's database, nodeIP:nodePort is the key.
+	// We have separate slice for TCP and UDP nodePort load-balancers.
+	// We will nodeIP separately later.
+	nodeportServices := make(map[string][]int32)
+
+	// Go through the k8s services and populate 'clusterServices' and
+	// 'nodeportServices'
+	for _, serviceInterface := range services {
+		service, ok := serviceInterface.(*kapi.Service)
+		if !ok {
+			logrus.Errorf("Spurious object in syncServices: %v",
+				serviceInterface)
+			continue
+		}
+
+		if service.Spec.Type != kapi.ServiceTypeClusterIP ||
+			service.Spec.Type != kapi.ServiceTypeNodePort {
+			continue
+		}
+
+		if service.Spec.ClusterIP == "" {
+			continue
+		}
+
+		for _, svcPort := range service.Spec.Ports {
+			protocol := svcPort.Protocol
+			if protocol == "" || (protocol != TCP && protocol != UDP) {
+				protocol = TCP
+			}
+
+			if service.Spec.Type == kapi.ServiceTypeNodePort {
+				port := svcPort.NodePort
+				if protocol == TCP {
+					nodeportServices[TCP] = append(nodeportServices[TCP], port)
+				} else {
+					nodeportServices[UDP] = append(nodeportServices[UDP], port)
+				}
+			}
+
+			if svcPort.Port == 0 {
+				continue
+			}
+
+			key := fmt.Sprintf("%s:%s", service.Spec.ClusterIP, svcPort.Port)
+			if protocol == TCP {
+				clusterServices[TCP] = append(clusterServices[TCP], key)
+			} else {
+				clusterServices[UDP] = append(clusterServices[UDP], key)
+			}
+		}
+	}
+
+	// Get OVN's current cluster load-balancer VIPs and delete them if they
+	// are stale.
+	for _, protocol := range []string{TCP, UDP} {
+		loadBalancer := ovn.getLoadBalancer(kapi.Protocol(protocol))
+		loadBalancerVIPS := ovn.getLoadBalancerVIPS(loadBalancer)
+		if loadBalancerVIPS == nil {
+			continue
+		}
+
+		for vip := range loadBalancerVIPS {
+			if !stringSliceMembership(clusterServices[protocol], vip) {
+				ovn.deleteLoadBalancerVIP(loadBalancer, vip)
+			}
+		}
+	}
+
+	// For each gateway, remove any VIP that does not exist in
+	// 'nodeportServices'.
+	gateways, err := ovn.getOvnGateways()
+	if err != nil {
+		logrus.Errorf("failed to get ovn gateways. Not syncing nodeport")
+		return
+	}
+
+	for _, gateway := range gateways {
+		physicalIP := ovn.getGatewayPhysicalIP(gateway)
+		if physicalIP == "" {
+			logrus.Errorf("physical gateway %s does not have physical ip",
+				gateway)
+			continue
+		}
+
+		for _, protocol := range []string{TCP, UDP} {
+			loadBalancer := ovn.getGatewayLoadBalancer(gateway, protocol)
+			if loadBalancer == "" {
+				logrus.Errorf("physical gateway %s does not have "+
+					"load_balancer", gateway)
+				continue
+			}
+
+			loadBalancerVIPS := ovn.getLoadBalancerVIPS(loadBalancer)
+			if loadBalancerVIPS == nil {
+				continue
+			}
+
+			for vip := range loadBalancerVIPS {
+				vipAndPort := strings.Split(vip, ":")
```

Comment:
You can replace about 10 lines here with net.SplitHostPort() I think...

Reply:
The SplitHostPort returns the port as a string too. So a conversion to int was needed. But, looks like we don't have to store ports as ints in the first place. So I used SplitHostPort and modified the datastructure to have ports stored as string.